xmtp_mls/src/groups/validated_commit_v2.rs

```diff
@@ -0,0 +1,465 @@
+    MlsCredential(#[from] BasicCredentialError),
+    #[error(transparent)]
+    ProtoDecode(#[from] prost::DecodeError),
+    #[error(transparent)]
+    InstallationDiff(#[from] InstallationDiffError),
+}
+
+#[derive(Debug, Clone, PartialEq, Hash)]
+pub(crate) struct CommitParticipant {
+    pub inbox_id: String,
+    pub installation_id: Vec<u8>,
+    pub is_creator: bool,
+    // TODO: Add is_admin
+}
+
+impl TryFrom<&LeafNode> for CommitParticipant {
+    type Error = CommitValidationError;
+
+    fn try_from(leaf_node: &LeafNode) -> Result<Self, Self::Error> {
+        Ok(Self {
+            inbox_id: inbox_id_from_credential(leaf_node.credential())?,
+            installation_id: leaf_node.signature_key().as_slice().to_vec(),
+            is_creator: false,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct Inbox {
+    pub inbox_id: String,
+    pub is_creator: bool,
+    // TODO: add is_admin support
+    // pub is_admin: bool,
+}
+
+/**
+ * A [`ValidatedCommit`] is a summary of changes coming from a MLS commit, after all of our validation rules have been applied
+ *
+ * Commit Validation Rules:
+ * 1. If the `sequence_id` for an inbox has changed, it can only increase
+ * 2. The client must create an expected diff of installations added and removed based on the difference between the current
+ * [`GroupMembership`] and the [`GroupMembership`] found in the [`StagedCommit`]
+ * 3. Installations may only be added or removed in the commit if they were added/removed in the expected diff
+ * 4. For updates (either updating a path or via an Update Proposal) clients must verify that the `installation_id` is
+ * present in the [`AssociationState`] for the `inbox_id` presented in the credential at the `to_sequence_id` found in the
+ * new [`GroupMembership`].
+ * 5. All proposals in a commit must come from the same installation
+ */
+#[derive(Debug, Clone)]
+pub struct ValidatedCommit {
+    pub actor: CommitParticipant,
+    pub inboxes_added: Vec<Inbox>,
+    pub inboxes_removed: Vec<Inbox>,
+}
+
+impl ValidatedCommit {
+    pub async fn from_staged_commit<'client, ApiClient: XmtpMlsClient + XmtpIdentityClient>(
+        conn: &'client DbConnection<'client>,
+        staged_commit: &StagedCommit,
+        openmls_group: &OpenMlsGroup,
+        client: &'client Client<ApiClient>,
+    ) -> Result<Self, CommitValidationError> {
+        // Get the actor who created the commit.
+        // Because we don't allow for multiple actors in a commit, this will error if two proposals come from different authors.
+        let actor = extract_actor(staged_commit, openmls_group)?;
+
+        // Get the expected diff of installations added and removed based on the difference between the current
+        // group membership and the new group membership.
+        // Also gets back the added and removed inbox ids from the expected diff
+        let (new_group_membership, expected_installation_diff, inboxes_added, inboxes_removed) =
```

Comment:
Does it make sense to create a struct for this? Could still get the individual variables with struct desugaring

Reply:
Ya. You're probably right. I'll change that up